consensus/hotstuff/validator/validator.go

```diff
@@ -46,13 +47,20 @@ func (v *Validator) ValidateQC(qc *flow.QuorumCertificate, block *model.Block) e
 
 	// Retrieve full Identities of all legitimate consensus participants and the Identities of the qc's signers
 	// IdentityList returned by hotstuff.Committee contains only legitimate consensus participants for the specified block (must have positive stake)
+	// TODO: filter.Any can be removed
 	allParticipants, err := v.committee.Identities(block.BlockID, filter.Any)
```

Comment:
When validating QC, we need to:
1. decode the signer IDs
2. check if the QC represents enough stakes
3. check if the QC's signature is correct.

If step 1 is failed, then step 2, 3 can be skipped. 

There is some data needed for both step 2 and 3, which is the signer identities, but it's not needed by step 1. But why I'm querying all the participants here before doing step 1?

It's a trade off. 

We could technically query all participants' IDs by view here from the committee module, who could cache all signer IDs with canonical order by view for each epoch. If step 1 failed, then we could skip step 2 and 3. But in the happy path where the validation will pass, where it need to query identities anyway for step 2 and 3. 

So rather, we just query all the identities here, which will be used by step 1,2 and 3. We simplifies the committee API by wasting one query to the committee module in the unhappy case of step 1 fails.

Reply:
👍 very convincing reasoning